### ISSUE TITLE
fan sets to full speed after cooling down

### ISSUE THREAD

**Issue by twoseascharlie** [neg2led/cm4io-fan#3]
I've built this (sudo dkms install cm4io-fan/0.1.0-2) on 32bit raspi and it seems to work well except for 1 issue. After it has been cooling and ramping down to slower and slower speeds, it kicks back on at full speed. Until the temp starts to rise again, then the driver appears to "kick on" and the fan speed slows down and then speeds up a bit until the pi is cooled off, and then this cycle repeats. I would expect the fan to completely shut off, or to remain at a low speed.

My setup:
CM4 with wifi + 4GB ram Lite With IO Board
pi@raspi-nas1:~ $ uname -r
5.10.49-v7l_charlie1+

my confit.txt:
dtoverlay=cm4io-fan,minrpm=300,maxrpm=1700,midtemp=50000,maxtemp=55000

My Fan:
Noctua NF-P12 redux-1700 PWM
(https://www.amazon.com/gp/product/B07CG2PGY6/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1)
Noctua NA-FC1, 4-Pin PWM Fan Controller
(https://www.amazon.com/gp/product/B072M2HKSN/ref=ppx_yo_dt_b_asin_title_o08_s00?ie=UTF8&psc=1)

I have tried connecting the fan directly to the pwm pins on the IO boards, and have tried with the noctua fan controller with attached sata power. I see the same behavior either way.

I notice that if I don't load the driver and just use these 2 commands to directly control the fan, it works as I would expect:
Turn fan on full speed: sudo i2cset -y 10 0x2f 0x30 0xff
Turn fan off: sudo i2cset -y 10 0x2f 0x30 0x00

So perhaps in your driver there should be some setting that after the fan has ramped down to just set the 0x30 register to 0 so the fan stays off?

**Comment by neggles** [comment]
Sorry for the slow reply here! GitHub didn't notify me of a new issue being opened, I must have something set up wrong...

First thing I notice here - you have `maxtemp` set _way_ too low. Only having a five degree difference between `midtemp` and `maxtemp` is going to cause all kinds of fan-speed hunting; set it to at _least_ `60000` (60 degrees C), preferably somewhere in the range between 65000 and 75000. There's no risk of damaging the Pi or reducing its lifespan at those temperatures, and it'll prevent the sudden ramp when you go from just under `midtemp` to over `maxtemp` in the space of a few seconds.

The other part of the behaviour you're seeing has to do with the NF-P12 redux's low minimum RPM, and a couple of details about the EMC2301's hardware. If you take a look at page 4 of the datasheet (which you can [download from the product page](https://www.microchip.com/en-us/product/EMC2301)) you might see the problem - the tachometer input range only goes down to 480RPM. Below that, you _might_ get a valid reading, or you might not.

If the chip detects an RPM that's too low - e.g. when it's set to 300, and it loses track of the actual RPM - it assumes the fan has failed, sets the PWM output to max speed, and throws an alarm. The driver currently doesn't clear the alarm until the temperature passes one of the setpoints again, so it sits there on max speed until that happens.

Long story short, you're seeing a combination of setting your `midtemp` and `maxtemp` too close together, and setting your `minrpm` below the chip's minimum tach input. 

Try changing your config.txt line to this, and you shouldn't have this issue anymore:
```
dtoverlay=cm4io-fan,minrpm=480,maxrpm=1800,midtemp=50000,maxtemp=70000
```

**Comment by twoseascharlie** [comment]
Thanks for all the info. I can see how that behavior makes sense. If I had some failing fan condition, try and turn it on full speed as a catchall to avoid overheating problems.

I have been trying various configurations, but I am still seeing the same behavior. When the fan is running and as it is cooling off and approaching midtemp, its speed slows down as expected. But then it still kicks on full speed eventually.
I've tried setting higher and higher minrpm values to try to keep the fan spinning above the chips min rpm of 480.

dtoverlay=cm4io-fan,minrpm=800,maxrpm=1800,midtemp=50000,maxtemp=70000.

For noise and for power usage, I would prefer to have it on at a low speed, or just off completely if nothing is running warm. So I wouldn't really even want to try any higher values for minrpm.

Perhaps this is just a quirk with this Noctua fan then since it operates at lower rpms?

I've written a quick and dirty user space daemon that works for my needs, so unless anyone else starts reporting issues with your driver similar to this, it's probably not a big deal.

**Comment by neggles** [comment]
Interesting. I had a look through the code again, and I've found something that could be the issue.

The EMC2301 has four different PWM frequencies it can output; the default is _supposed_ to be 26KHz, which is within the 25-28KHz "acceptable range" in the intel spec for four-wire fans, but in some older datasheets the default is listed as 2.441KHz - if the frequency is set too low, it could be upsetting the Noctua's electronics at lower duty cycles.

The code (as it currently is) doesn't set the PWM frequency register, so it's possible the chip is coming up in a lower frequency mode; there are also some other parameters that can be adjusted (target RPM tolerance and PID gains), and it can be set to read RPM as higher than it actually is by changing the pole count.

It also lacks a "fan off" mode / threshold, which is something I'd personally like to have as well, so I might be making some improvements to this (when I can find the time...)

There are two other things you can do; if you create `/etc/modprobe.d/emc2301.conf` with these contents:
```
options emc2301 register_cdev=0
```
This will stop the driver from registering itself for automatic control, but the hwmon device will still be created, so you can control the thing by echoing an RPM setpoint to `/sys/class/hwmonX/fan1_target`

The second thing you can do is, make DKMS uninstall and forget about the driver (the version might be 0.1.0-2, if you used install.sh):
```
dkms uninstall cm4io-fan/0.1.0 --all
dkms remove cm4io-fan/0.1.0 --all
```
Then go into `/usr/src/cm4io-fan-0.1.0/emc2301/` and edit `emc2301.c`, adding these two lines at the top:

```
#undef dev_dbg
#define dev_dbg dev_info
```
Make DKMS reinstall the module:
```
dkms add cm4io-fan/0.1.0
dkms install cm4io-fan/0.1.0
```
Reboot, and you should get a bunch of extra debug info in `dmesg` when various things happen (fan RPM alerts, RPM setpoints etc).

I could make this a dynamic module parameter, but it'd make the code a bit... _messy_ since I'm not the greatest of C programmers and `dev_dbg` usually gets optimized out at compile time, so changing the function it calls at runtime is a bit... weird.

I have a bunch of Noctua PWM fans, but no reduxes, or I'd test this myself...

**Comment by ok7mjm** [comment]
Faced similar issue: when target rpm is set below 1000 no pwm signal is present on fan control pin (100% pwm), input rpm is maxed. Everything works fine if minrpm is set above 1000 rpm.
Acording EMC2301 dasheet: 

> RNGx[1:0]: Range. Sets the minimum fan speed measured and reported

Looks like it's 0b01 by default, that means "1000 RPM minimum, TACH count multiplier = 2", I see in logs  `emc2301 10-002f: Writing 0xAB to 0x32`.
Changing range fixes the problem.

**Comment by twoseascharlie** [comment]
I can confirm that setting a slightly higher minrpm than I was testing before works as I would expect, i.e.:

this allows the fan to stay at a lower speed as it cools down:
dtoverlay=cm4io-fan,minrpm=850,maxrpm=1800,midtemp=50000,maxtemp=70000

this does not, and the fan ramps up to full speed after it cools:
dtoverlay=cm4io-fan,minrpm=800,maxrpm=1800,midtemp=50000,maxtemp=70000

I experimented with changing the RNGx bits ok7mjm mentioned to try and see if I could get it to support lower speeds, but was unsuccessful.

I tried adding these 2 lines to 0 out the RNG bits and set to 500 RPM minimum:
fan_config_reg_val &= ~(1<<6);
fan_config_reg_val &= ~(1<<5);

inside the if(enable) block here:
https://github.com/neg2led/cm4io-fan/blob/master/emc2301/emc2301.c#L337

Looking at the dmesg logs, it looks like the bits were set, but I still am unable to set a lower fan speed without it ramping up. Maybe it has something to do with the TACH count multiplier that is also controlled by those 2 bits? I don't have a lot of time to look at this myself right now. But it would be nice if something like that were possible for lower speed fans.

**Comment by ok7mjm** [comment]
Just to be sure. Did you also change `FAN_RPM_FACTOR * 2` to `FAN_RPM_FACTOR * 1`? This magic number is TACH count multiplier which depends on range setting.

**Comment by neggles** [comment]
It's not _quite_ that simple, unfortunately... but I think I have a fix here. Just changing RNGx alone doesn't actually solve the problem, it's just changing how the chip _reports_ RPM. Currently, the driver sets these parameters in the config reg (the `0xab` value, `0b10101011`)

| Name   | ENAGx   | RNGx   | EDGx  | UDGx   |
|--------|---------|--------|-------|--------|
| Descr. | Enable  | Range  | Edges | Update |
| Bits   | 1       | 01     | 01    | 011    |
| Value  | Enabled | mult=1 | 2 pulses/rev  | 400ms  |

Most PC fans are two-pole fans that output two tachometer pulses per actual physical fan revolution - I've not come across anything other than some old AIO watercooler pumps that output anything else. 

These settings configure the EMC2301 it to count two pulses on the tacho line as one revolution (EDGx setting) and applies a x1 multiplier to that value to get actual RPM, but has the unfortunate side-effect of being unable to read below 1000RPM.

Here's an actual fix, setting the register to `0x83`:

| Name   | ENAGx   | RNGx   | EDGx  | UDGx   |
|--------|---------|--------|-------|--------|
| Descr. | Enable  | Range  | Edges | Update |
| Bits   | 1       | 00     | 00    | 011    |
| Value  | Enabled | mult=0.5 | 1 pulse/rev | 400ms  |

This reads one pulse as one revolution, which (by itself) would only serve to make the RPM reading double what it actually is, but it _also_ applies a 0.5 multiplier to that value to counteract that. Amusingly this is _supposed_ to be the default register value for those bits, according to the datasheet (all bits initialized to 0).

If you change

```
	if (enable) {
        fan_config_reg_val |= (1<<7);
	} else {
        fan_config_reg_val &= ~(1<<7);
	}
```
to
```
	if (enable) {
        // enable fan control
        fan_config_reg_val |= (1<<7);
        // configure tach for low-RPM fans
        fan_config_reg_val &= ~(0xf << 3);
	} else {
        fan_config_reg_val &= ~(1<<7);
	}
```
that should put it in this mode, in theory...

Edit: Note that this does also reduce the maximum RPM that you can read from the device, and it doesn't alter any of the other code that calculates RPM values, so things might be a bit off?

**Comment by twoseascharlie** [comment]
@ok7mjm
I had not changed the (FAN_RPM_FACTOR * 2) to (FAN_RPM_FACTOR * 1).

I can confirm that after changing that and with my initial changes still in place:
fan_config_reg_val &= ~(1<<6);
fan_config_reg_val &= ~(1<<5);

the fan works as I would expect. It speeds up as it gets warmer and slows down as it gets cooler, without ever ramping up. I have been able to successfully set the minrpm down to 500:
dtoverlay=cm4io-fan,minrpm=500,maxrpm=1800,midtemp=50000,maxtemp=70000

@neg2led
I tried using your change:
// configure tach for low-RPM fans
fan_config_reg_val &= ~(0xf << 3);

and the fan never speeds up, it stays at a low speed no matter what the temp is. I tried this with and without changing (FAN_RPM_FACTOR * 2) to (FAN_RPM_FACTOR * 1).

I know you mentioned there may be more that needs to be updated in the code, so just passing on the info I have gathered.

I don't have a higher speed fan to test with, but based on my needs, I would say that zeroing out the 2 range bits along with changing the FAN_RPM_FACTOR multiplier calculation works great for how I would expect a driver like this to behave.

Appreciate the active discussion on this issue :)

**Comment by neggles** [comment]
OK, great! The more I look at this datasheet and run calculations / compare against what I'm seeing in my own setup, the less certain I am about my posted solution, so I'm not surprised it doesn't work right. I am still not entirely clear on what the `EDGx` bits actually do to the TACH count register values; changing it from 5 edges per rotation to 3 edges per rotation _should_ double the RPM reading and halve the COUNT value, all else being equal, but doesn't seem to actually do that, so I'm missing something there.

One thing I _am_ sure of is, changing the `EDGx` value away from the default 5/rotation setting breaks the "simplified" RPM calculation formula that this module uses - that's what I'm referring to when I say other changes would be needed - but it's probably not worth the effort/overhead to change that out, especially when it doesn't seem to actually help anything.

Anyway. Support for fans that spin below 1000rpm is definitely something this should have; I can think of a couple different ways to implement it as a dynamic option:

- Automatically switch minimum RPM range based on `midrpm` setting, maybe set to the low range if `midrpm` is less than 2000?
- Add a dtparam (`rpmrange=0` / `rpmrange=1`) to choose between the two
- add a module parameter to choose between the two

But it looks like even at `RNGx = 0b00` the maximum reportable RPM is well north of 20,000, so I'm not sure this even needs to be dynamic, I can probably just change it to default to 500rpm as a minimum.

Thoughts?

**Comment by neggles** [comment]
Yeah, let's just not complicate things, this happily reads out fan RPMs in excess of 18,000 so I'm gonna call that good enough.

@twoseascharlie / @ok7mjm can you test the contents of #4 before I merge it in, just to be safe?

**Comment by ok7mjm** [comment]
Yep, works for me as expected. Thank you for the fast response!

**Comment by neggles** [comment]
Great, I'm going to call that good then - for now, at least. Maybe at some point I'll have the time to re-write this in the style of the existing EMC2103 (not a confusing naming scheme at all, Microchip/SMSC...) hwmon module and expose more configuration options...

I suppose we can close this now? 😝

**Comment by twoseascharlie** [comment]
Great thanks! :tada:

Can confirm it works as expected for me as well. I like the idea of just keeping it simple, especially if you can confirm it works for higher rated rpms. 

> Maybe at some point I'll have the time to re-write this in the style of the existing EMC2103 (not a confusing naming scheme at all, Microchip/SMSC...) 

Haha yes, when I first started looking for a driver for this, digging through the kernel config menu...

**Comment by neggles** [comment]
I attached a server fan that's rated at 18,000RPM and that read out just fine as ~17,900RPM, so... I'm going to call that good enough. There _are_ some small fans that can hit north of 22,000, but I've not found any that are 4-pin (and actually available to buy) - if someone comes along with one of those, I'll make it dynamic 😝
